zebra-network/src/address_book.rs

```diff
@@ -5,54 +5,73 @@ use std::{
     collections::{BTreeSet, HashMap},
     iter::Extend,
     net::SocketAddr,
+    sync::Arc,
 };
 
 use chrono::{DateTime, Utc};
 use tracing::Span;
 
-use crate::{
-    constants,
-    types::{MetaAddr, PeerServices},
-};
+use crate::{constants, types::MetaAddr, PeerConnectionState};
 
 /// A database of peers, their advertised services, and information on when they
 /// were last seen.
 #[derive(Debug)]
 pub struct AddressBook {
-    by_addr: HashMap<SocketAddr, (DateTime<Utc>, PeerServices)>,
-    by_time: BTreeSet<MetaAddr>,
+    /// Maps each peer address to its matching `MetaAddr`
+    by_addr: HashMap<SocketAddr, Arc<MetaAddr>>,
+
+    /// Stores peer addresses, in state then reconnection order.
+    ///
+    /// This order should not be used for reconnection attempts: use
+    /// [`reconnection_peers`] instead.
+    by_state: BTreeSet<Arc<MetaAddr>>,
```

Comment:
Remove the `Arc` and just copy the `MetaAddr`

Reply:
>     * See if there's a Rust in-memory database-like type that can do lookups on multiple distinct keys, and combined keys, and ensure key uniqueness

there doesn't seem to be, maybe we should just use `HashSet<SocketAddr, MetaAddr>` and see what the performance is like

Reply:
The performance is acceptable, even with ~4,000 failed addresses in the list on mainnet.